tests/Type/WebMozartAssert/data/impossible-check-eq-not-eq.php

```diff
@@ -0,0 +1,39 @@
+<?php declare(strict_types=1);
+
+namespace WebmozartAssertImpossibleCheckEqNotEq;
+
+use DateTimeInterface;
+use stdClass;
+use Webmozart\Assert\Assert;
+
+function sameInstancesAreAlwaysEqual(stdClass $a, stdClass $b): void
+{
+	Assert::eq($a, $a); // will always evaluate to true
+	Assert::notEq($b, $b); // will always evaluate to false
+}
+
+function instancesOfTheSameTypeAreNotIdenticalButCouldBeEqual(stdClass $a, stdClass $b, stdClass $c, stdClass $d): void
+{
+	Assert::eq($a, $b);
+	Assert::notEq($c, $d);
+
+	Assert::eq(createStdClass(), createStdClass());
+	Assert::notEq(createStdClass(), createStdClass());
+}
+
+function instancesOfDifferentTypesAreNeverEqual(stdClass $a, DateTimeInterface $b, stdClass $c, DateTimeInterface $d, stdClass $e, stdClass $f, stdClass $g, stdClass $h): void
+{
+	Assert::eq($a, $b); // ignored, should evaluate to false?
+	Assert::notEq($c, $d); // ignored, should evaluate to true?
```

Comment:
These are fine. PHPStan currently doesn't implement the messy semantics from here: https://www.php.net/manual/en/types.comparisons.php ("Loose comparisons with ==" table).

Reply:
Yes, it's not currently handled at all.

I just pushed some changes to phpstan/phpstan-src dev-master, I'm pretty sure it will change some results here after it lands in phpstan/phpstan (once this finishes https://github.com/phpstan/phpstan-src/actions/runs/1957754903). Feel free to wait a bit and require ^1.4.9 here.

Reply:
great. Just for my understanding, is it easy to explain why the changes you made (e.g. https://github.com/phpstan/phpstan-src/commit/98c4d84ec983fb36992119d4c393d7192a2a99b6) are in MutatingScope and not in TypeSpecifier?

Reply:
aah, I assume you're kind of building expressions there for the TypeSpecifier to specify later. Makes sense then